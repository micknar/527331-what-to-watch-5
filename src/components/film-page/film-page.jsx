import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {Redirect} from 'react-router-dom';
import FilmsList from "../films-list/films-list";
import PageHeaderLogo from "../page-header-logo/page-header-logo";
import UserBlock from "../user-block/user-block";
import FilmPageOverview from "../film-page-overview/film-page-overview";
import FilmPageDetails from "../film-page-details/film-page-details";
import FilmPageReviews from "../film-page-reviews/film-page-reviews";
import PageFooter from "../page-footer/page-footer";

const FilmPageNav = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

class FilmPage extends PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      filmTab: FilmPageNav.OVERVIEW,
    };
  }

  render() {
    const {films, onFilmCardClick} = this.props;
    const {id, name, posterImage, backgroundImage, description, rating, runTime, director, starring, genre, released, reviews} = films[0];

    const getFilmInfo = () => {
      switch (this.state.filmTab) {
        case FilmPageNav.OVERVIEW:
          return (
            <FilmPageOverview
              description={description}
              rating={rating}
              director={director}
              starring={starring}
            />
          );
        case FilmPageNav.DETAILS:
          return (
            <FilmPageDetails
              runTime={runTime}
              genre={genre}
              director={director}
              starring={starring}
              released={released}
            />
          );
        case FilmPageNav.REVIEWS:
          return (
            <FilmPageReviews
              reviews={reviews}
            />
          );
      }

      return <Redirect to="/" />;
    };

    return (
      <>
        <section className="movie-card movie-card--full" data-id={id}>
          <div className="movie-card__hero">
            <div className="movie-card__bg">
              <img src={backgroundImage} alt={name} />
            </div>

            <h1 className="visually-hidden">WTW</h1>

            <header className="page-header movie-card__head">
              <PageHeaderLogo />
              <UserBlock />
            </header>

            <div className="movie-card__wrap">
              <div className="movie-card__desc">
                <h2 className="movie-card__title">{name}</h2>
                <p className="movie-card__meta">
                  <span className="movie-card__genre">{genre}</span>
                  <span className="movie-card__year">{released}</span>
                </p>

                <div className="movie-card__buttons">
                  <button className="btn btn--play movie-card__button" type="button">
                    <svg viewBox="0 0 19 19" width="19" height="19">
                      <use xlinkHref="#play-s"></use>
                    </svg>
                    <span>Play</span>
                  </button>
                  <button className="btn btn--list movie-card__button" type="button">
                    <svg viewBox="0 0 19 20" width="19" height="20">
                      <use xlinkHref="#add"></use>
                    </svg>
                    <span>My list</span>
                  </button>
                  <a href="add-review.html" className="btn movie-card__button">Add review</a>
                </div>
              </div>
            </div>
          </div>

          <div className="movie-card__wrap movie-card__translate-top">
            <div className="movie-card__info">
              <div className="movie-card__poster movie-card__poster--big">
                <img src={posterImage} alt={name} width="218" height="327" />
              </div>

              <div className="movie-card__desc">
                <nav className="movie-nav movie-card__nav">
                  <ul className="movie-nav__list">
                    <li className="movie-nav__item movie-nav__item--active">
                      <a href="#" className="movie-nav__link"
                        onClick={() => {
                          this.setState({
                            filmTab: FilmPageNav.OVERVIEW,
                          });
                        }}>Overview</a>
                    </li>

                    <li className="movie-nav__item">
                      <a href="#" className="movie-nav__link"
                        onClick={() => {
                          this.setState({
                            filmTab: FilmPageNav.DETAILS,
                          });
                        }}>Details</a>
                    </li>

                    <li className="movie-nav__item">
                      <a href="#" className="movie-nav__link"
                        onClick={() => {
                          this.setState({
                            filmTab: FilmPageNav.REVIEWS,
                          });
                        }}>Reviews</a>
                    </li>
                  </ul>
                </nav>

                {getFilmInfo()}
              </div>
            </div>
          </div>
        </section>

        <div className="page-content">
          <section className="catalog catalog--like-this">
            <h2 className="catalog__title">More like this</h2>

            <FilmsList
              films={films.slice(0, 4)}
              onFilmCardClick={onFilmCardClick}
            />
          </section>

          <PageFooter />
        </div>
      </>
    );
  }
}

FilmPage.propTypes = {
  films: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    posterImage: PropTypes.string.isRequired,
    backgroundImage: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    rating: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    released: PropTypes.number.isRequired,

    runTime: PropTypes.shape({
      hours: PropTypes.number.isRequired,
      minutes: PropTypes.number.isRequired,
    }),

    reviews: PropTypes.arrayOf(PropTypes.shape({
      user: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
      }),
      rating: PropTypes.string.isRequired,
      comment: PropTypes.string.isRequired,
      date: PropTypes.shape({
        day: PropTypes.number.isRequired,
        month: PropTypes.string.isRequired,
        monthNumber: PropTypes.number.isRequired,
        year: PropTypes.number.isRequired,
      }),
    }))
  })),
  onFilmCardClick: PropTypes.func.isRequired,
};

export default FilmPage;
