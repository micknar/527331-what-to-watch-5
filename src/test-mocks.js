import {getDate} from "./utils";

export const noop = () => {};

export const films = [
  {
    id: 1,
    name: `Bronson`,
    posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    backgroundImage: `/img/the-grand-budapest-hotel-bg.jpg`,
    backgroundColor: `#4aa2e5`,
    videoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
    rating: 8.9,
    ratingMark: `Very good`,
    scoresCount: 240,
    director: `Nicolas Winding Refn`,
    starring: [`Bill Murray`, `Saoirse Ronan`],
    runTime: {
      hours: 2,
      minutes: 40,
    },
    genre: `Crime`,
    released: 1996,
    isFavorite: false,
  },
];

export const comments = [
  {
    id: 1,
    user: {
      id: 19,
      name: `Christina`,
    },
    rating: 2.4,
    comment: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam`,
    date: getDate(`2020-10-09T13:38:44.769Z`),
  },
  {
    id: 2,
    user: {
      id: 14,
      name: `Corey`,
    },
    rating: 3.2,
    comment: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat`,
    date: getDate(`2020-10-27T13:38:44.769Z`),
  },
];

export const userInfo = {
  'id': 1,
  'email': `Oliver.conner@gmail.com`,
  'name': `Oliver.conner`,
  'avatar_url': `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/9.jpg`,
};

export const filmsFromServer = [{
  id: 1,
  name: `Bronson`,
  // eslint-disable-next-line camelcase
  poster_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  // eslint-disable-next-line camelcase
  preview_image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  // eslint-disable-next-line camelcase
  background_image: `/img/the-grand-budapest-hotel-bg.jpg`,
  // eslint-disable-next-line camelcase
  background_color: `#4aa2e5`,
  // eslint-disable-next-line camelcase
  video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  // eslint-disable-next-line camelcase
  preview_video_link: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
  rating: 8.9,
  // eslint-disable-next-line camelcase
  scores_count: 240,
  director: `Nicolas Winding Refn`,
  starring: [`Bill Murray`, `Saoirse Ronan`],
  // eslint-disable-next-line camelcase
  run_time: 160,
  genre: `Crime`,
  released: 1996,
  // eslint-disable-next-line camelcase
  is_favorite: false,
}];

export const commentFromServer = [
  {
    id: 1,
    user: {
      id: 19,
      name: `Christina`,
    },
    rating: 2.4,
    comment: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam`,
    date: `2020-10-09T13:38:44.769Z`,
  },
  {
    id: 2,
    user: {
      id: 14,
      name: `Corey`,
    },
    rating: 3.2,
    comment: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat`,
    date: `2020-10-27T13:38:44.769Z`,
  },
];
