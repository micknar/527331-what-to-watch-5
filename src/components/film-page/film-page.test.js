import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from "react-router-dom";
import {FilmPage} from "./film-page";
import {noop, films} from "../../test-mocks";

jest.mock(`../films-list/films-list`, () => `FilmsList`);
jest.mock(`../page-header-logo/page-header-logo`, () => `PageHeaderLogo`);
jest.mock(`../user-block/user-block`, () => `UserBlock`);
jest.mock(`../page-footer/page-footer`, () => `PageFooter`);
jest.mock(`../film-page-tabs/film-page-tabs`, () => `FilmPageTabs`);
jest.mock(`../add-to-favorite-btn/add-to-favorite-btn`, () => `AddToFavoriteBtn`);
jest.mock(`../loading-page/loading-page`, () => `LoadingPage`);
jest.mock(`../not-found-page/not-found-page`, () => `NotFoundPage`);

describe(`Should FilmPage render correctly`, () => {
  it(`With loading success`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <FilmPage
              films={films}
              onCardClick={noop}
              currentFilmId={1}
              currentFilm={films[0]}
              getFilm={noop}
              isFilmLoading={false}
            />
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`When data is loading`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <FilmPage
              films={films}
              onCardClick={noop}
              currentFilmId={1}
              currentFilm={films[0]}
              getFilm={noop}
              isFilmLoading={true}
            />
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`When film id is wrong`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <FilmPage
              films={films}
              onCardClick={noop}
              currentFilmId={99999999}
              currentFilm={films[0]}
              getFilm={noop}
              isFilmLoading={false}
            />
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
