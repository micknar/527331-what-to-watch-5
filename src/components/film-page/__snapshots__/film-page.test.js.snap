// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FilmPage render correctly When data is loading 1`] = `<LoadingPage />`;

exports[`Should FilmPage render correctly When film id is wrong 1`] = `<NotFoundPage />`;

exports[`Should FilmPage render correctly With loading success 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
    data-id={1}
    style={
      Object {
        "backgroundColor": "#4aa2e5",
      }
    }
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Bronson"
          src="/img/the-grand-budapest-hotel-bg.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <PageHeaderLogo />
        <UserBlock />
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Bronson
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Crime
            </span>
            <span
              className="movie-card__year"
            >
              1996
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/player/1"
              onClick={[Function]}
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <AddToFavoriteBtn
              id={1}
              isFavorite={false}
            />
            <a
              className="btn movie-card__button"
              href="/films/1/review"
              onClick={[Function]}
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="Bronson"
            height="327"
            src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg"
            width="218"
          />
        </div>
        <FilmPageTabs
          activeTab="Overview"
          film={
            Object {
              "backgroundColor": "#4aa2e5",
              "backgroundImage": "/img/the-grand-budapest-hotel-bg.jpg",
              "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
              "director": "Nicolas Winding Refn",
              "genre": "Crime",
              "id": 1,
              "isFavorite": false,
              "name": "Bronson",
              "posterImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg",
              "previewImage": "https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg",
              "previewVideoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
              "rating": 8.9,
              "ratingMark": "Very good",
              "released": 1996,
              "runTime": Object {
                "hours": 2,
                "minutes": 40,
              },
              "scoresCount": 240,
              "starring": Array [
                "Bill Murray",
                "Saoirse Ronan",
              ],
              "videoLink": "https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm",
            }
          }
          onActiveTabClick={[Function]}
        />
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <FilmsList
        activeCard={-1}
        films={Array []}
        onCardClick={[Function]}
        onCardMouseOut={[Function]}
        onCardMouseOver={[Function]}
      />
    </section>
    <PageFooter />
  </div>,
]
`;
